### PR TITLE
Refactor events for MVVM pattern

### PR DESCRIPTION
**Pull request by willwull**
Followed [this tutorial](https://medium.com/flawless-app-stories/how-to-use-a-model-view-viewmodel-architecture-for-ios-46963c67be1b) and tried to rewrite the event controllers according to the MVVM pattern.

### CODE REVIEW

File: Studs/Events/ViewModels/EventCellViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+//  Created by Willy Liu on 2019-02-22.
+//  Copyright © 2019 Studieresan. All rights reserved.
+//
+
+import Foundation
+
+struct EventCellViewModel {
+    let month: String
+    let day: String
+    let companyName: String?
```
Comment (mictab) [inline Studs/Events/ViewModels/EventCellViewModel.swift]: Can a company not have a name?

File: Studs/Events/ViewModels/EventsViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        // create an EventTableCellViewModel for each event
+        for (index, category) in events.enumerated() {
+            for event in category {
+                guard let date = event.getDate() else {
+                    fatalError("Couldn't format date of event")
+                }
+
+                // Display month and date
+                let dateFormatter = DateFormatter()
```
Comment (mictab) [inline Studs/Events/ViewModels/EventsViewModel.swift]: This one could be cached. No need to create a new DateFormatter for each cell.

File: Studs/Events/ViewModels/EventsViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    private let disposeBag = DisposeBag()
+
+    // What should happen when the cellVMs are changed
+    var reloadTableViewClosure: (() -> Void)?
+
+    // MARK: Public methods
+
+    func fetchData() {
+        // Fetch the events
+        Http.fetchAllEvents().subscribe(onNext: { events in
```
Comment (mictab) [inline Studs/Events/ViewModels/EventsViewModel.swift]: You probably want a weak reference to `self` here. Imagine what happens when the network request starts and the user leaves the view.

File: Studs/Events/ViewModels/EventCellViewModel.swift (hunk 1 above)
Comment (willwull) [reply Studs/Events/ViewModels/EventCellViewModel.swift]: Right good point. The `Event` struct looks like this:

```swift
struct Event: Decodable {
    let id: String
    let companyName: String?
    // ...
```

So if I set `companyName` as non-optional, line 108 in EventsViewModel would give me a warning:
```swift
let viewModel = EventCellViewModel(month: month, day: day, companyName: event.companyName)
```

**Comment by willwull** [comment]
@mictab Had to add two `guard` statements in the data fetching. Not sure if this is the correct way to use weak `self`.

Also, wasn't sure whether you meant having `dateFormatter` outside the loops, or as a property on the class itself.

File: Studs/Events/ViewModels/EventCellViewModel.swift (hunk 1 above)
Comment (mictab) [reply Studs/Events/ViewModels/EventCellViewModel.swift]: It should probably not be null there either then. It doesn't make any sense for a company to not have a name.

File: Studs/Events/ViewModels/EventsViewModel.swift [hunk 4]
```diff
@@ -35,14 +35,19 @@ final class EventsViewModel {
 
     func fetchData() {
         // Fetch the events
-        Http.fetchAllEvents().subscribe(onNext: { events in
+        Http.fetchAllEvents().subscribe(onNext: { [weak self] events in
             // Sort events, earliest first
```
Comment (mictab) [inline Studs/Events/ViewModels/EventsViewModel.swift:41]: You can do `guard let self = self else { return }` here and then you don't need to do `self?`. You might have to write it as
> guard let \`self\` = self else { return }

**Review by mictab — COMMENTED** [review]
Regarding the DateFormatter: either way is fine.

**Comment by willwull** [comment]
`guard let self = self else { return }` did the trick!

**Review by mictab — APPROVED** [review]

**Pull request merged**